### ISSUE TITLE
Cannot edit post with mention to first post of other discussions

### ISSUE THREAD

**Issue by clarkwinkelmann** [flarum/core#1770]
## Bug Report

**Current Behavior**
When editing a post that contains a mention to another post that is the first post in its discussion, the text formatter assumes that post will either not be in the store, or be in the store and have a discussion relationship. However this isn't always true.

**Steps to Reproduce**
1. Write a reply that mentions the first post of the discussion. That post must be from another user or old enough so it doesn't appear on your own profile page without loading more posts
2. Navigate to the forum homepage (important) and refresh the page
3. Navigate to your user profile by using the link the header (or by going through a discussion unrelated to those tests) in the same window without a page reload
4. Click edit on the last post that includes a mention
5. `TypeError: e.discussion(...).id is not a function` error is triggered and the editor never appears

The line in question is part of the mentions extension here

https://github.com/flarum/mentions/blob/e95386610b4362a81ba13b86dd226895abd5af9a/js/src/forum/utils/textFormatter.js#L19

**Other way to Reproduce**
1. Write a reply that mentions the first post from another discussion that is visible on the homepage (this requires copy-pasting as the editor does not offer it via the mention button)
2. Navigate to the forum homepage (important) and refresh the page
3. Navigate to the post you just created without a page reload
4. Click edit on that post
5. `TypeError: e.discussion(...).id is not a function` error is triggered and the editor never appears

**Expected Behavior**
The editor should open, there should be no errors.

**Screenshots**
![image](https://user-images.githubusercontent.com/5264300/56066033-d99d4e80-5d76-11e9-8b98-ce168998e43d.png)

![image](https://user-images.githubusercontent.com/5264300/56066022-cdb18c80-5d76-11e9-87de-c4abf4195ff4.png)

**Environment**
- Flarum version: beta 8 as well as the dev-master currently on Discuss
- Website URL: https://discuss.flarum.org/
- Webserver: not relevant
- Hosting environment: not relevant
- PHP version: not relevant
- Browser: Firefox 66, but probably not relevant

```
Output of "php flarum info", run this in terminal in your Flarum directory.
```

**Possible Solution**
The line https://github.com/flarum/mentions/blob/e95386610b4362a81ba13b86dd226895abd5af9a/js/src/forum/utils/textFormatter.js#L19 assumes that a post will always have a discussion relationship, however this is not the case.

If a post was loaded as the `firstPost` relationship of a discussion, then it doesn't have this relationship to the discussion. If you browse from the homepage, you put this record in the store. If you then navigate to a post that mentions this post and try to edit it, the javascript line will try to access post.discussion().id() and fail.

I'm not sure how much logic goes behind the mention system, but the fix might be as simple as checking if post.discussion() is false, in which case skip the `setAttribute()` call just like if the post itself was not in the store.

**Comment by tobscure** [comment]
Ideally we should always at least have a discussion _ID_ in the store. So instead of `post.discussion().id()`, would `post.data.relationships.discussion.data.id` work?

**Comment by clarkwinkelmann** [comment]
@tobscure in this situation, the post has no `post.data.relationships.discussion` at all. The API request could include `firstPost` (like now) and `firstPost.discussion` (not currently done), but that would add additional database queries and return the discussion two times in the response (well, the data is there once, but I suppose the performance penalty is there because we retrieved it twice)

Yes ideally loading the post as a child of a discussion should map the discussion as a relationship of the post, without having to include "2" times in the API response. But I really don't know how to achieve that. There's a danger of recursion depending on the implementation.

**Comment by tobscure** [comment]
Yep, perhaps tobscure/json-api needs some kind of inverse relationship functionality so that we can force the `firstPost`'s `discussion` relationship to be automatically populated without any performance overhead. I will think about this for tobscure/json-api-server which we will eventually adopt.
